### ISSUE TITLE
arrow transform generates buggy result when using .bind(this, 'arg')

### ISSUE THREAD

**Issue by lucmult** [lebab/lebab#287]
### Sample original source :
```js
function abc() {
  const callbacks = [];

  for (let i = 0; i < 10; i++ ) {
    const cb = function(idx) {
      console.log('idx: ' + idx);
    }.bind(this, i);  // <---- HERE using 

    callbacks.push(cb);
  }

  return callbacks;
};

const callbacks = abc();
for (let cb of callbacks) {
  cb();
}
```

### Transformed with this command
```bash
$ lebab bla.js --transform arrow > bla-transformed.js
```
### Result:

```js
function abc() {
  const callbacks = [];

  for (let i = 0; i < 10; i++ ) {
    const cb = idx => {
      console.log('idx: ' + idx);
    };

    callbacks.push(cb);
  }

  return callbacks;
};

const callbacks = abc();
for (let cb of callbacks) {
  cb();
}
```

Expected result:
```js
function abc() {
  const callbacks = [];

  for (let i = 0; i < 10; i++ ) {
    const cb = (idx => {
      console.log('idx: ' + idx);
    }).bind(null, i);  // <-- It needs to wrap in a () before calling .bind().

    callbacks.push(cb);
  }

  return callbacks;
```

**Comment by lucmult** [comment]
Example failing test:
lucmult/lebab/pull/1

I don't know how to fix it though. :-D

**Comment by uniibu** [comment]
ECMAScript 2015 spec: 
```Any reference to arguments, super, this, or new.target within an ArrowFunction must resolve to a binding in a lexically enclosing environment. Typically this will be the Function Environment of an immediately enclosing function.```

Meaning tachnically you cannot change the meaning/value of `this`. Some also say you cannot rebind an arrow function. However, you can still use bind in an àrrow function.


Using your example: 
```
function abc() {
  const callbacks = [];

  for (let i = 0; i < 10; i++ ) {
    const cb = function(idx) {
      console.log('idx: ' + idx);
    }.bind(null, i);  

    callbacks.push(cb);
  }

  return callbacks;
};

const callbacks = abc();
for (let cb of callbacks) {
  cb();
}

```

Becomes what you expected:
```
function abc() {
  const callbacks = [];

  for (let i = 0; i < 10; i++ ) {
    const cb = (idx => {
      console.log('idx: ' + idx);
    }).bind(null, i);  

    callbacks.push(cb);
  }

  return callbacks;
```

Test it out on https://lebab.unibtc.me/editor

**Comment by lucmult** [comment]
Hi uniibu,

The problem isn't with binding "this" but rather for the additional arguments.

That's why the expected solution has binds null instead of `this`.

Cheers.

**Comment by nene** [comment]
Thanks for reporting @lucmult, that's definitely a bug.

I've fixed it by not performing the arrow transform altogether when `.bind()` is also used for arguments.

As @uniibu pointed out, use of .bind() on arrow function is an odd thing as one can't change the `this` value. It might be that in your case, you'd like to replace `.bind(this, 1, 2, 3)` with `.bind(null, 1,2,3)`, but I'm not sure if that's really the way most people would like to handle this situation. I for one would rather consider rewriting this function in some other way.

So at least for start I fixed it by avoiding this transform in this case. Released in v3.0.4.

Note that in your example, you're actually not using `this` value inside the bound function at all, so the binding of `this` is kinda useless there. Because of that, Lebab 3.0.4 will simply transform the function and leave the .bind call in place. Only when you actually use `this` inside the function will Lebab consider whether the function happens to be bound.

I'm not really sure about this solution. Maybe I should also avoid transforming it when function contains no `this`. Maybe I should also produce a warning in this case. Feedback appreciated.

Most importantly though, Lebab won't screw up the code any more :)

**Comment by lucmult** [comment]
Hi @nene 

Thanks for the speedy resolution here. I don't have hard feelings either way (convert to bind(nulll, ...) or don't convert at all).  I agree with you that the important is to not generate that bug. :-)

Binding values to arguments is still a valid use for bind() and arrow functions. In the code that I was updating, it was actually using `this`:
https://chromium-review.googlesource.com/c/chromium/src/+/1493638/4/ui/file_manager/file_manager/foreground/js/ui/banners.js
https://chromium-review.googlesource.com/c/chromium/src/+/1493638/4/ui/file_manager/file_manager/foreground/js/ui/progress_center_panel.js

> ... Maybe I should also avoid transforming it when function contains no this. ... 

I for one, am happy to have functions without `this` converted as well.

Nevertheless, I'm happy to close this issue with the current solution.
